effect/shared/src/main/scala/scalaz/effect/IO.scala

```diff
@@ -0,0 +1,554 @@
+  }
+
+  final case class Fail[A](failure: Throwable) extends IO[A] {
+    override final def tag = Tags.Fail
+  }
+
+  final case class AsyncEffect[A](register: (Throwable \/ A => Unit) => AsyncReturn[A]) extends IO[A] {
+    override final def tag = Tags.AsyncEffect
+  }
+
+  final case class Attempt[A](value: IO[A]) extends IO[Throwable \/ A] {
+    override final def tag = Tags.Attempt
+  }
+
+  final case class Fork[A](value: IO[A], handler: Maybe[Throwable => IO[Unit]]) extends IO[Fiber[A]] {
+    override final def tag = Tags.Fork
+  }
+
+  final case class Race[A0, A1, A](left: IO[A0], right: IO[A1], finish: (A0, Fiber[A1]) \/ (A1, Fiber[A0]) => IO[A]) extends IO[A] {
+    override final def tag = Tags.Race
+  }
+
+  final case class Suspend[A](value: () => IO[A]) extends IO[A] {
+    override final def tag = Tags.Suspend
+  }
+
+  final case class Bracket[A, B](acquire: IO[A], release: (BracketResult[B], A) => IO[Unit], use: A => IO[B]) extends IO[B] {
+    override final def tag = Tags.Bracket
+  }
+
+  final case class Uninterruptible[A](io: IO[A]) extends IO[A] {
+    override final def tag = Tags.Uninterruptible
+  }
+
+  final case class Sleep(duration: Duration) extends IO[Unit] {
+    override final def tag = Tags.Sleep
+  }
+
+  final case class Supervise[A](value: IO[A], error: Throwable) extends IO[A] {
+    override final def tag = Tags.Supervise
+  }
+
+  /**
+   * Lifts a strictly evaluated value into the `IO` monad.
+   */
+  final def now[A](a: A): IO[A] = Strict(a)
+
+  /**
+   * Lifts a non-strictly evaluated value into the `IO` monad. Do not use this
+   * function to capture effectful code. The result is undefined but may
+   * include runtime errors.
+   */
+  final def point[A](a: => A): IO[A] = Point(() => a)
+
+  /**
+   * Raises the specified error. The moral equivalent of `throw` for pure code.
+   */
+  final def fail[A](t: Throwable): IO[A] = Fail(t)
+
+  /**
+   * Strictly-evaluated unit lifted into the `IO` monad.
+   */
+  final val unit: IO[Unit] = now(())
+
+  /**
+   * Sleeps for the specified duration. This is always asynchronous.
+   */
+  final def sleep(duration: Duration): IO[Unit] = Sleep(duration)
+
+  /**
+   * Supervises the specified action, which ensures that any actions directly
+   * forked by the action are killed with the specified error upon the action's
+   * own termination.
+   */
+  final def supervise[A](io: IO[A], error: Throwable): IO[A] = Supervise(io, error)
+
+  /**
+   * Flattens a nested action.
+   */
+  final def flatten[A](io: IO[IO[A]]): IO[A] = io.flatMap(a => a)
+
+  /**
+   * Lazily produces an `IO` value whose construction may have actional costs
+   * that should be be deferred until evaluation.
+   *
+   * Do not use this method to effectfully construct `IO` values. The results
+   * will be undefined and most likely involve the physical explosion of your
+   * computer in a heap of rubble.
+   */
+  final def suspend[A](io: => IO[A]): IO[A] = Suspend(() => io)
```

Comment:
I'm a little confused by the expected behavior - should lazy mean both deferred and memoized like it is for scala's lazy vals or just deferred?  This currently behaviors just as deferred - if you have, something e 1like the following, wouldn't you expect the output to be 0, then 1?  Instead, subsequent invocations, re-construct the IO, so it's not memoized; only deferred (i.e. not the same lazy as scala, which has both properties).

```
object MyApp extends SafeApp {
  lazy val intGetter = IO.suspend {
    val atomicInt = new AtomicInteger()
    IO.sync {
      atomicInt.getAndIncrement
    }
  }
  for {
   intOne <- intGetter
    _ <- putStrLn(intOne)
   intTwo <- intGetter
    _ <- putStrLn(intTwo)
  } yield ()
}
```

Reply:
Unfortunately this code is invalid. You are performing an effect inside the block passed to `suspend`, so the semantics of the code are undefined.

`suspend` is designed only to defer the (pure) construction of an `IO`. This is needed in some cases to ensure lazy evaluation in the presence of runtime fusion.

Other literature (incorrectly) encourages the use of `suspend` for performing effects (such as memoization).